pkg/action/failback.go

```diff
@@ -0,0 +1,213 @@
+				migrationSchedule.Spec.Suspend = &suspend
+			}
+			msg := fmt.Sprintf("Suspending migration schedule %s before doing failback operation", migrationSchedule.Name)
+			log.ActionLog(action).Infof(msg)
+			_, err := storkops.Instance().UpdateMigrationSchedule(migrationSchedule)
+			if err != nil {
+				log.ActionLog(action).Errorf("Error suspending migration schedule %s: %v", migrationSchedule.Name, err)
+			}
+			action.Status.Status = storkv1.ActionStatusSuccessful
+			ac.updateAction(action)
+		} else {
+			// Fail the action failback if latest successful migration is older than latestMigrationThresholdTime
+			// And migrationschedule is in suspended state
+			log.ActionLog(action).Infof("The latest migration %s is not in threshold range", latestMigration.Name)
+			msg := fmt.Sprintf("Failing failback operation as the latest migration %s is not in threshold range", latestMigration.Name)
+			log.ActionLog(action).Errorf(msg)
+			action.Status.Status = storkv1.ActionStatusFailed
+			action.Status.Reason = msg
+			ac.recorder.Event(action,
+				v1.EventTypeWarning,
+				string(storkv1.ActionStatusFailed),
+				msg)
+			ac.updateAction(action)
+		}
+	}
+}
+
+func (ac *ActionController) performDeactivateFailBackStage(action *storkv1.Action) {
+	if action.Status.Status == storkv1.ActionStatusSuccessful {
+		action.Status.Stage = storkv1.ActionStageLastMileMigration
+		action.Status.Status = storkv1.ActionStatusInitial
+		action.Status.FinishTimestamp = metav1.Now()
+		ac.updateAction(action)
+		return
+	} else if action.Status.Status == storkv1.ActionStatusFailed {
+		action.Status.Stage = storkv1.ActionStageFinal
+		action.Status.FinishTimestamp = metav1.Now()
+		ac.updateAction(action)
+		return
+	}
+
+	migrationSchedule, err := storkops.Instance().GetMigrationSchedule(action.Spec.ActionParameter.FailbackParameter.MigrationScheduleReference, action.Namespace)
+	if err != nil {
+		msg := fmt.Sprintf("error fetching migrationschedule %s/%s for failback", action.Namespace, action.Spec.ActionParameter.FailbackParameter.MigrationScheduleReference)
+		log.ActionLog(action).Errorf(msg)
+		action.Status.Status = storkv1.ActionStatusFailed
+		action.Status.Reason = msg
+		ac.recorder.Event(action,
+			v1.EventTypeWarning,
+			string(storkv1.ActionStatusFailed),
+			msg)
+		ac.updateAction(action)
+		return
+	}
+
+	namespaces, err := utils.GetMergedNamespacesWithLabelSelector(migrationSchedule.Spec.Template.Spec.Namespaces, migrationSchedule.Spec.Template.Spec.NamespaceSelectors)
+	if err != nil {
+		msg := fmt.Sprintf("Failed to get list of namespaces from migrationschedule %s", migrationSchedule.Name)
+		log.ActionLog(action).Infof(msg)
+		ac.recorder.Event(action,
+			v1.EventTypeWarning,
+			string(storkv1.ActionStatusScheduled),
+			msg)
+		ac.updateAction(action)
+		return
+	}
+
+	for _, ns := range namespaces {
+		err := resourceutils.ScaleReplicasReturningError(ns, true, true, ac.printFunc(action, "ScaleReplicas"), ac.config)
```

Comment:
I am thinking of another case. What if user performs a failover from C1->C2, and then adds new apps in C2 before doing a failover from C2->C1. In this case the new apps will not have a migrationReplicas annotation. I don't think this case is handled by current updateReplicas method.
Also another point if the originalReplicaCount is 0, do we still want to update the replicaAnnotation which will then be carried forward to the source cluster C1 ??

Reply:
I see your point. To handle this case, one way can be, we can reuse the same flow of checking what is getting activated and then deactivate the corresponding app (similar to failover). I will do that in another PR.